lib/pacemaker/pcmk_sched_ordering.c

```diff
@@ -1208,54 +1209,67 @@ find_actions_by_task(pe_resource_t *rsc, const char *original_key)
 {
     GList *then_actions = NULL;
-    pe_action_t *then_action = NULL;
-    uint32_t flags;
+    uint32_t flags = pe_order_none;
 
-    CRM_CHECK(rsc != NULL, return);
-    CRM_CHECK(order != NULL, return);
+    CRM_CHECK((rsc != NULL) && (order != NULL), return);
 
     flags = order->flags;
-    then_action = order->rh_action;
-    crm_trace("Applying ordering constraint %d (then: %s)", order->id, rsc->id);
+    pe_rsc_trace(rsc, "Applying ordering %d for 'then' resource %s",
+                 order->id, rsc->id);
 
-    if (then_action != NULL) {
-        then_actions = g_list_prepend(NULL, then_action);
+    if (order->rh_action != NULL) {
+        then_actions = g_list_prepend(NULL, order->rh_action);
 
-    } else if (rsc != NULL) {
+    } else {
         then_actions = find_actions_by_task(rsc, order->rh_action_task);
     }
 
     if (then_actions == NULL) {
-        pe_rsc_trace(rsc,
-                     "Ignoring constraint %d: then (%s for %s) not found",
+        pe_rsc_trace(rsc, "Ignoring ordering %d: no %s actions found for %s",
                      order->id, order->rh_action_task, rsc->id);
         return;
     }
 
     if ((first_action != NULL) && (first_action->rsc == rsc)
         && pcmk_is_set(first_action->flags, pe_action_dangle)) {
 
-        pe_rsc_trace(rsc, "Detected dangling operation %s -> %s",
-                     first_action->uuid, order->rh_action_task);
+        pe_rsc_trace(rsc,
+                     "Detected dangling migration ordering (%s then %s %s)",
+                     first_action->uuid, order->rh_action_task, rsc->id);
         pe__clear_order_flags(flags, pe_order_implies_then);
     }
 
-    for (GList *gIter = then_actions; gIter != NULL; gIter = gIter->next) {
-        pe_action_t *then_action_iter = (pe_action_t *) gIter->data;
+    if ((first_action == NULL) && !pcmk_is_set(flags, pe_order_implies_then)) {
+        pe_rsc_debug(rsc,
+                     "Ignoring ordering %d for %s: No first action found",
+                     order->id, rsc->id);
+        g_list_free(then_actions);
+        return;
+    }
+
+    for (GList *iter = then_actions; iter != NULL; iter = iter->next) {
+        pe_action_t *then_action_iter = (pe_action_t *) iter->data;
 
         if (first_action != NULL) {
             order_actions(first_action, then_action_iter, flags);
-
-        } else if (pcmk_is_set(flags, pe_order_implies_then)) {
-            pe__clear_action_flags(then_action_iter, pe_action_runnable);
-            crm_warn("Unrunnable %s %#.6x", then_action_iter->uuid, flags);
         } else {
-            crm_warn("neither %s %#.6x", then_action_iter->uuid, flags);
+            pe__clear_action_flags(then_action_iter, pe_action_runnable);
+            crm_warn("%s of %s is unrunnable because there is no %s of %s "
+                     "to order it after", then_action_iter->task, rsc->id,
+                     order->lh_action_task, order->lh_rsc->id);
```

Comment:
This is probably correct. It should work fine for the call from `pcmk__apply_ordering()`. I'm not as sure about when this function is called from `rsc_order_first()`, where we have a list of `first_actions` that are based on the `first_rsc` argument (rather than based on `order->lh_rsc` directly).

This is likely close enough

Reply:
On second look, the only place `rsc_order_first()` gets called actually DOES use the `lh_rsc` of its `order` argument... so its first argument is redundant, but improving it is likely out of scope here.